src/QsCompiler/SyntaxProcessor/StatementVerification.fs

```diff
@@ -212,116 +144,85 @@ let private VerifyBinding tryBuildDeclaration (qsSym, (rhsType, rhsEx, rhsRange)
 /// Generates an InvalidUseOfTypeParameterizedObject error if the given variable type contains external type parameters
 /// (i.e. type parameters that do no belong to the parent callable associated with the given symbol tracker).
 /// Returns the pushed declaration as Some, if the declaration was successfully added to given symbol tracker, and None otherwise.
-let private TryAddDeclaration isMutable
-                              (symbols: SymbolTracker)
-                              (name: string, location, localQdep)
-                              (rhsType: ResolvedType, rhsEx, rhsRange)
-                              =
-    let t, tpErr =
-        if rhsType.isTypeParametrized symbols.Parent
-           || IsTypeParamRecursion (symbols.Parent, symbols.DefinedTypeParameters) rhsEx then
-            InvalidType |> ResolvedType.New,
-            [|
-                rhsRange |> QsCompilerDiagnostic.Error(ErrorCode.InvalidUseOfTypeParameterizedObject, [])
-            |]
-        else
-            rhsType, [||]
-
+let private TryAddDeclaration isMutable (symbols: SymbolTracker) (name: string, location, localQdep) rhsType =
+    let t, tpErr = rhsType, [||]
     let decl = LocalVariableDeclaration<_>.New isMutable (location, name, t, localQdep)
     let added, errs = symbols.TryAddVariableDeclartion decl
-    (if added then Some decl else None), errs |> Array.append tpErr
+    (if added then Some decl else None), Array.append tpErr errs
 
 /// Given a Q# symbol, as well as the expression on the right hand side that is assigned to it,
 /// resolves and verifies the assignment using VerifyBinding and the resolution context.
 /// Pushes all determined variable declarations into the current scope of the symbol tracker,
 /// generating the corresponding error(s) if a symbol with the same name is already visible on that scope and/or the symbol is not a valid variable name.
 /// Builds the corresponding Q# let- or mutable-statement (depending on the given kind) at the given location,
 /// and returns it along with an array of all generated diagnostics.
-let private NewBinding kind comments (location: QsLocation) context (qsSym: QsSymbol, qsExpr: QsExpression) =
-    let rhs, rhsErrs = Verify context qsExpr
-
-    let symTuple, varDeclarations, errs =
-        let isMutable =
-            match kind with
-            | MutableBinding -> true
-            | ImmutableBinding -> false
-
-        let localQdep = rhs.InferredInformation.HasLocalQuantumDependency
+let private NewBinding kind comments (location: QsLocation) context (symbol, expr) =
+    let expr, diagnostics = resolveExpr context expr
 
+    let symbolTuple, varDeclarations, bindingDiagnostics =
         let addDeclaration (name, range) =
-            TryAddDeclaration isMutable context.Symbols (name, (Value location.Offset, range), localQdep)
+            TryAddDeclaration
+                (kind = MutableBinding)
```

Comment:
I am looking for the ImmutableBinding and why this is always Mutable, but I don't see it - what am I missing? Is this somehow needed to make the second pass updating the syntax tree to include type information? How would you know then which symbols are mutably bound and which ones aren't in the final AST?

Reply:
Hmm, I think there is some confusion caused by F# syntax here. Note that the original definition of `isMutable`:

```fsharp
let isMutable =
    match kind with
    | MutableBinding -> true
    | ImmutableBinding -> false
```

is equivalent to the expression `kind = MutableBinding` (`=` as in "is equal to", not assignment). Thus I simply replaced `isMutable` with the expression `(kind = MutableBinding)`, so the behavior is the same in both cases. But it does look like a named argument...

Reply:
Yes, indeed, looking at the code and having worked all in C# lately I am now expecting == for equality comparisons...